### PR TITLE
#76 working workaround proposal

### PR DESCRIPTION
**Pull request by tapy-lappy**

### CODE REVIEW

File: ClientWeb/Evolution/Components/gene.component.ts [hunk 1]
```diff
@@ -54,6 +55,16 @@ export class GeneComponent extends BaseGeneComponent implements OnInit, OnDestro
...
     }
 
+    manuallyCastTo(source: any): Site[] {
+        //Workaround: when we do deep copy of some item with ObjectHelper.deepCopy(), it stores the structure of tree of properties.
+        //Workaround: BUT it also looses the typing, so the structure is the same, but it's anonimous type. And if original type has
+        //Workaround: some computable property, it's lost in anonimous type. It's exactly what happens with Site.name computable prop.
+        //Done: So, to fix this problem created interface ManuallyCast which contains this method manuallyCast and the solution is
+        //Done: RE-CREATING original type by using data from anonimous type that we have. It's workaround but I have doubts the problem
+        //Done: maybe resolved any different way:
+        return source.map((item:Site) => new Site(item.site, item.isMutated));
```
Comment (AndreyAttr) [inline ClientWeb/Evolution/Components/gene.component.ts:65]: Recreating of item(s) is bad approach. It means we must implement this ManuallyCast<T> interface everywhere where we need to use deep copy. Also what if there were several deep copies, as Site, Gene etc.? Then with this approach we must implement several interfaces: `ManuallyCast<Site>, ManuallyCast<Gene>`.

**Review by AndreyAttr — CHANGES_REQUESTED** [review]
@tapy-lappy ,
Ok, I undestood the problem. I have to refuse your workaround-solution, it's quick fix and we have to find other way to resolve problem

File: ClientWeb/Evolution/Components/gene.component.ts [hunk 2]
```diff
@@ -125,7 +138,7 @@ export class GeneComponent extends BaseGeneComponent implements OnInit, OnDestro
     }
     private switchMutation(mutationSites?: Site[]) : (enabled:boolean)=>void{
         if(mutationSites)
-            this.gene.mutationSites = mutationSites.map(item => new Site(item.site, item.isMutated));   //Workaround
+            this.gene.mutationSites = this.manuallyCastTo(mutationSites);   //workaround
```
Comment (tapy-lappy) [inline ClientWeb/Evolution/Components/gene.component.ts:141]: Here we had the same workaround. I just moved it into one place - in interface

File: ClientWeb/Evolution/Components/gene.component.ts (hunk 1 above)
Comment (tapy-lappy) [reply ClientWeb/Evolution/Components/gene.component.ts:65]: I tried to find more acceptable way. Of couse we need to do it into ObjectHelper.deepCopy() but whatever I tried there were no way to save typings.

File: ClientWeb/Evolution/Components/gene-editor.component.ts [hunk 3]
```diff
@@ -80,14 +81,14 @@ export class GeneEditorComponent extends BaseGeneComponent implements OnInit, On
...
                 //Remark: also sets this.currentChosenSite which is also bounded(so do not need to clean up it manually):
-                this.editorForm.reset()
+                this.editorForm.reset();
         });
     }
 
     add(){
-        //TODO: Angular/TS allows to do copies simplier - find it
-        //Note: Using local LET variable to prevent setting all the this.gene properties to NULL when we reset the form
-        let gene = new Gene(this.gene.name, [this.currentChosenSite], this.gene.description);
```
Comment (tapy-lappy) [inline ClientWeb/Evolution/Components/gene-editor.component.ts:90]: Here we had this workaround.

**Review by tapy-lappy — COMMENTED** [review]
So 2 places with the same quick fix code, as you can see.

File: ClientWeb/Evolution/Components/gene.component.ts (hunk 2 above)
Comment (AndreyAttr) [reply ClientWeb/Evolution/Components/gene.component.ts:141]: sad

File: ClientWeb/Evolution/Components/gene-editor.component.ts (hunk 3 above)
Comment (AndreyAttr) [reply ClientWeb/Evolution/Components/gene-editor.component.ts:90]: I see. Yep, this is bad approach. deepCopy() is developed especially to change behaviour like that, not to use 'recreate instance' pattern as **workaround** everywhere, but have specialized method in one place which do all the work.

File: ClientWeb/Evolution/Components/gene.component.ts (hunk 1 above)
Comment (AndreyAttr) [reply ClientWeb/Evolution/Components/gene.component.ts:65]: @tapy-lappy,
Ok, I got it. I need to do some fixes in #76. But this approach is also unacceptable, so let it hold on till I find the way to resolve problem. When I do I just close this pull request.
Thank you so much for your investigation, lad.

**Comment by AndreyAttr** [comment]
@tapy-lappy , I fixed #76 to support typing(ans so computable properties & instance members as methods). Also provided there `cast()` method to get back typing(manually) from FormGroup/FormArray etc. Reactive Forms component. So #76 is not blocker for your #65 anymore. I reject this pull request. Please update from **angular** branch.

File: ClientWeb/Evolution/Components/gene.component.ts (hunk 1 above)
Comment (AndreyAttr) [reply ClientWeb/Evolution/Components/gene.component.ts:65]: @tapy-lappy , done. You may check 2 my last commits on #76. Also found we lose typings when got back model from Reactive Form control, so now there is `cast()` method to make back-typing.
> It's workaround but I have doubts the problem
>  //Done: maybe resolved any different way:

Fortunatelly, yes. Thx you found all these stuff.

**Pull request closed without merging**